### ISSUE TITLE
Assertion failed: Permission denied (..\..\..\src\tcp_connecter.cpp:284)

### ISSUE THREAD

**Issue by bryanedds** [zeromq/fszmq#84]
I'm building an asynchronous REQ socket like so -

    open System
    open System.Text
    open fszmq
    
    let [<Literal>] private TcpBindAddress = "tcp://*:"
    let [<Literal>] private TcpConnectAddress = "tcp://localhost:"
    let [<Literal>] private BalancerForemanSocketDomainAddress = "5560"
    let [<Literal>] private BalancerForemanSocketBindAddress = TcpBindAddress + BalancerForemanSocketDomainAddress
    let [<Literal>] private BalancerForemanSocketConnectAddress = TcpConnectAddress + BalancerForemanSocketDomainAddress
    let [<Literal>] private WorkOrdersToExecute = 256
    let [<Literal>] private WorkOrderContent = "!desreveR"

    let private executeWork workNumber context =
        use bfSocket = Context.req context
        Socket.setOption bfSocket (fszmq.ZMQ.IDENTITY, string <| Guid.NewGuid ())
        Socket.connect bfSocket BalancerForemanSocketConnectAddress
        let workOrderContentEncoded = Encoding.UTF8.GetBytes WorkOrderContent
        Socket.send bfSocket workOrderContentEncoded
        ignore <| Socket.recv bfSocket
        if Socket.recvMore bfSocket then failwith "Programming error: Expecting single recv to be sufficient in Balancer's executeWork."

    let private executeWorkAsync workNumber context =
        async { return executeWork workNumber context }

    let [<EntryPoint>] main _ =

        // build zeromq context
        use context = new Context ()

        // execute several work orders in parallel
        seq { for i in 1 .. WorkOrdersToExecute do yield executeWorkAsync i context }
        |> Async.Parallel
        |> Async.RunSynchronously
        |> ignore
    
        // result in success
        0

When this program is running on an unbound connection (IE - no clients, and therefore blocking many threads on the Socket.recv call), the program crashes with no .NET stack trace with -

    Assertion failed: Permission denied (..\..\..\src\tcp_connecter.cpp:284)

The program is typically run from Visual Studio Ultimate 2013 in debug mode, but also exhibits the behavior when run stand-alone. The operating environment is Windows 8.1, x64. fszmq version is 4.0.6.

We have not been able to determine what internally is going wrong, and see no place where we violate ZMQ's admittedly quite implicit invariants.

**Comment by pblasucci** [comment]
Thanks for reporting this. Please try using the latest (pre-release) version:

https://www.nuget.org/packages/fszmq/4.0.8-prerelease

If this doesn't address the issue, I will investigate further.

**Comment by bryanedds** [comment]
Yes, just checked and it still happens in 4.0.8-prerelease :/

I almost suspect a bug inside ZeroMQ, but who knows. Please look into this for us - we really appreciate your time and support!

**Comment by pblasucci** [comment]
Rotten. I was afraid that might happen. And I also suspect this might be a bug in `libzmq.dll`. Regardless, I look into this later tonight.

**Comment by bryanedds** [comment]
EDIT - I just now updated the above code to remove some external dependencies that I accidentally left in. Should compile and run fine on your end now!

**Comment by pblasucci** [comment]
So, after a bit of testing, I'm not able to reproduce this issue. I wonder if there's some other piece of environmental configuration that I'm missing (e.g. security, version of MSVCRT.dll, et cetera)?

**Comment by pblasucci** [comment]
After experimenting with things for a few hours, I was unable to reproduce the error. So I started looking at the native source code from where the error originated. 

_(Aside: I have an updated NuGet package queued up... just waiting on the "official" updates to be merged upstream.)_ 

Now I'm almost certain the issue is environmental (possibly related to security). The code in question is signaling `WSAEACCES` when it tries to call `getsockopt` (a WinSock function) as part of making a TCP connection. This is very likely due to firewall restrictions, insufficient permissions, an invalid socket address, or some other environmental concern. Could you perhaps try running the code on someone else's machine? If not, perhaps we can compare OS-level settings and figure out if that's what's causing it. 

Additionally, perhaps we could inquire with the ZeroMQ mailing list (zeromq-dev@lists.zeromq.org)? This doesn't appear to be an issue in the managed portion of the codebase, so I'm going to close the issue. But please let me know how I can help you further debug this situation.

**Comment by bryanedds** [comment]
So, I'm trying to build the zeromq dll myself to do my own debugging, and also get a hold of the 4.0.5 and 4.1.0RC versions that don't seem to be available as builds. Can you give me some advice or a link on how to build this thing for windows? I've tried using cmake, make, and all that, but nothing yields any particular progress, likely due to my complete lack of familiarity with these build tools.

**Comment by pblasucci** [comment]
Generally, I fork the relevant ZMQ repositories on GitHub. Then I clone locally and build. The repos are as follows:

+ [4.0 branch](https://github.com/zeromq/zeromq4-x.git)
+ [4.1 branch](https://github.com/zeromq/zeromq4-1.git)

As for the actual building process, if you look in the path `builds\msvc` you'll find solution files for various versions of Visual Studio. These always seem to work quite well for me.

Also, I've just release the updated NuGet package (with the 4.0.7-prerelease bug fixes and an updates `libzmq.dll`).

**Comment by bryanedds** [comment]
Ha, that's weird. When I opened the .sln from the sources I got from the zeromq website, all the project files couldn't be loaded because they were missing. I suppose in that version I was supposed to run CMake to generate them. But fortunately, when pulling them from github, the project files are already there :)

Thanks again, Paul!

**Comment by bryanedds** [comment]
Hm... missing winsock2.h... Any idea what version of Windows SDK I should install? I'm running Windows 8.1.

**Comment by bryanedds** [comment]
FYI, updated to your latest, still getting the same issue. I think at this point I'm going to move forward under the assumption of it being a simple permissions problem.

Is there a systematic way to compare our Window system permissions settings?

**Comment by pblasucci** [comment]
Heh... you've been busy.   :-)

As far as SDKs go, I always install whatever the latest one is for the OS.

With respect to environmental settings, I'm not 100% certain how to proceed. I think the very first thing to check is which version(s) of MSVCRT.dll is(are) installed on your machine. For instance, I have several copies on my PC: some have no version, others are versions 2, 6, and 7 respectively (n.b. the versions are NOT in the file name... one has to check the file properties). Meanwhile, I will research ways to compare systems permissions (I suspect this will be complicated by Active Directory... but here's to hoping).

**Comment by bryanedds** [comment]
So, I feel horribly embarrassed, but -

The issue was caused by a bloatware version of McAfee AV that came on the new PC I just received at work. I'm suppose what was happening was that McAfee was seeing some 'suspicious' utilization of my ports by ZeroMQ, and decided, after some period of time, to deny permissions to this sort of activity, thus resulting in said assertion. All, of course, without giving me any actual user notifications.

The solution, of course, was to disable McAfee's firewall protection. Ultimately, due to frustration and a bit of paranoia, I uninstalled the entire thing.

Sorry to waste so much of your time on such a silly issue!
